### ISSUE TITLE
[sw tests,dv]  Passing arguments from cfg hjson file to bazel

### ISSUE THREAD

**Issue by sha-ron** [lowRISC/opentitan#13430]
Need to transfer bazel_opts to a global integer such that we can add options from chip_ast_sim_cfg.hjson.
For example, we need to transfer MANUFACTURER_HOOKS_DIR.
@sriyerg @tjaychen  FYI.

**Comment by sha-ron** [comment]
Thank you @timothytrippel

**Comment by ariel9678** [comment]
@timothytrippel 
Thank you for  #13693. it partially solved the problem.
Currently the sim.mk defines in “bazel_label”  the test location and test name as :  “//sw/device/tests/closed_source:closed_chip_sw_alert_test_sim_dv”
When running test from closed source the test label should be “@manufacturer_test_hooks//:closed_chip_sw_alert_test_sim_dv”

**Comment by timothytrippel** [comment]
yes, #13612 should fix this issue once merged, working on cleaning it up and getting that submitted soon.

**Comment by timothytrippel** [comment]
#13863 should finally resolve this.

**Comment by timothytrippel** [comment]
Since this is taking a little longer than expected, I wanted to share more details on what resolving this issue involves / the progress that has been made for transparency.

First, while this issue primarily states "passing Bazel arguments from HJSON cfg file to Bazel", it also implies (states in a follow up comment) that the goal run tests defined in the "manufacturere_test_hooks" Bazel repository in DV sim (e.g., “@manufacturer_test_hooks//:closed_chip_sw_alert_test_sim_dv”). So this issue is really two issues in one. To solve both, multiple changes must be made to our dvsim/Bazel infrastructure:

- [x] add support to dvsim for "sw_build_opts" or flags that can be passed directly to the Bazel build system (#13693)
- [x] Use Bazel labels to define SW images in the dvsim HJSON config files. Until recently, we were referring to SW images in these config files with paths to where the target lives in our tree / where it will live after it gets build (i.e. we relied on these being the same). This does not work for the `manufacturer_test_hooks` Bazel repo as it lives in an entirely different Bazel repo at a (potentially) differerent system path. Hence we needed to let Bazel handle filesystem paths and just pass it target labels. (#13612)
- [ ] Make `opentitan_functest` artifact labels queryable. Currently, the test artifacts that are generated by the `opentitan_functest` (i.e., flash, rom, and otp images) are [copied into the testbench directory](https://cs.opensource.google/opentitan/opentitan/+/master:hw/dv/tools/dvsim/sim.mk;drc=ecd843648f2b0af9082b514dd0eabb0cb0eda8f1;l=100) after being built. This works when the bazel target string matches the built artifacts files names. However if there was ever a mismatch this would not work. To solve this we can query bazel for the targets that are being built and just copy those file directly. However the  `opentitan_functest` artifact labels need to be made queryable through a refactor of the `opentitan_functest` and supporting Bazel macros. (in progress)
- [x] Enable manufacturer tests to run in DV sim. Once the above changes are made we can enable an example manufacture test in the `chip_sim_cfg.hjson` file to demonstrate these capabilities (#13863)

**Comment by ariel9678** [comment]
Thank you Tim.
Trying to implement it, we are still getting the following errors:
ERROR: Skipping '@manufacturer_test_hooks//:closed_chip_sw_alert_test_sim_dv': no such target '@manufacturer_test_hooks//:closed_chip_sw_alert_test_sim_dv': target 'closed_chip_sw_alert_test_sim_dv' not declared in package '' defined by /home/users/arcohen/_bazel_arcohen/4018e373332d5ea1abead4c114838380/external/manufacturer_test_hooks/BUILD.bazel
WARNING: Target pattern parsing failed.
ERROR: no such target '@manufacturer_test_hooks//:closed_chip_sw_alert_test_sim_dv': target 'closed_chip_sw_alert_test_sim_dv' not declared in package '' defined by /home/users/arcohen/_bazel_arcohen/4018e373332d5ea1abead4c114838380/external/manufacturer_test_hooks/BUILD.bazel

May be it is because we are working outside the opentitan repository?

**Comment by ariel9678** [comment]
Tim, 
The MANUFACTURER_HOOKS_DIR is set to the location of the local closed source.
It seems that when adding new test in local folder, although the test added to the local BUILD.bazel ,if the test name was not defined in open source BUILD.bazel ($REPO_TOP/sw/device/tests/closed_source/BUILD.bazel ) the build will fail, with target test not found.

**Comment by ariel9678** [comment]
Trying to add new test under closed source I see the following behavior:
(All test executions run with dv_sim script)
1.      Running the example_test (appear in closed & open source directories) with MANUFACTURER_HOOKS_DIR fleg - test pass.
2.      Added error to example_test in my local closed source directory – the build failed with compilation error
3.      Added new_test to BUILD.bazel in my local closed source directory.
4.      Running new_test from my closed source directory with MANUFACTURER_HOOKS_DIR – new_test target was not found (build failed)
5.      Replaced example_test name with new_test name in BUILD.bazel under $REPO_TOP/sw/device/tests/closed_source/ -  new test was build and run successfully from my local closed source.
	opentitan_functest(
              name = "new_test",
              srcs = ["example_test.c"],
              deps = [
              "@//sw/device/lib/testing/test_framework:ottf_main",
              ],
	)

-         When running directly from command line the build command produced by [sim.mk](http://sim.mk/)
(./bazelisk.sh build --action_env=MANUFACTURER_HOOKS_DIR=/nuvoton/verification/sw/device/tests/closed_source --define DISABLE_VERILATOR_BUILD=true @manufacturer_test_hooks//:new_test_sim_dv) the build pass successfully.
It seems that if target test name was not defined in open source BUILD.bazel ($REPO_TOP/sw/device/tests/closed_source/BUILD.bazel ) ,test build from local closed source directory will fail, with target not found.

**Comment by ariel9678** [comment]
- In chip_ast_sim_cfg.hjson defined:
   {
     name: new_test
     uvm_test_seq: ntil_chip_sw_alert_vseq
     sw_images: ["@manufacturer_test_hooks//:new_test:1"]
     en_run_modes: ["sw_test_mode_test_rom"]
     sw_build_opts: [" --action_env=MANUFACTURER_HOOKS_DIR=/nuvoton/verification/sw/device/tests/closed_source"]
   }
- new_test definition in BUILD.bazel in my local folder:
opentitan_functest(
   name = "new_test",
   srcs = ["new_test.c"],
   deps = [
       "@//sw/device/lib/base:mmio",
       "@//sw/device/lib/runtime:hart",
       "@//sw/device/lib/runtime:log",
       "@//sw/device/lib/testing:alert_handler_testutils",
       "@//sw/device/lib/dif:alert_handler",
       "@//sw/device/lib/dif:sensor_ctrl",
       "@//hw/top_earlgrey/ip/sensor_ctrl/data:sensor_ctrl_regs",
       "@//hw/top_earlgrey:alert_handler_regs",
       "@//sw/device/lib/testing/test_framework:ottf_main",
       "//ip:ast_regs",
       "//dif:ast",
   ],
)
- In BUILD.bazel under $REPO_TOP/sw/device/tests/closed_source/ , Replaced example_test name with new_test name :
opentitan_functest(
   name = "new_test",
   srcs = ["example_test.c"],
   deps = [
       "@//sw/device/lib/testing/test_framework:ottf_main",
   ],
)
Only when new_test name defined in BUILD.bazel under $REPO_TOP/sw/device/tests/closed_source/ - the  build and run of new_test in my local folder passed successfully

**Comment by ariel9678** [comment]
- In chip_ast_sim_cfg.hjson defined:
   {
     name: new_test
     uvm_test_seq: ntil_chip_sw_alert_vseq
     sw_images: ["@manufacturer_test_hooks//:new_test:1"]
     en_run_modes: ["sw_test_mode_test_rom"]
     sw_build_opts: [" --action_env=MANUFACTURER_HOOKS_DIR=/nuvoton/verification/sw/device/tests/closed_source"]
   }
- new_test definition in BUILD.bazel in my local folder:
opentitan_functest(
   name = "new_test",
   srcs = ["new_test.c"],
   deps = [
       "@//sw/device/lib/base:mmio",
       "@//sw/device/lib/runtime:hart",
       "@//sw/device/lib/runtime:log",
       "@//sw/device/lib/testing:alert_handler_testutils",
       "@//sw/device/lib/dif:alert_handler",
       "@//sw/device/lib/dif:sensor_ctrl",
       "@//hw/top_earlgrey/ip/sensor_ctrl/data:sensor_ctrl_regs",
       "@//hw/top_earlgrey:alert_handler_regs",
       "@//sw/device/lib/testing/test_framework:ottf_main",
       "//ip:ast_regs",
       "//dif:ast",
   ],
)
- In BUILD.bazel under $REPO_TOP/sw/device/tests/closed_source/ , Replaced example_test name with new_test name :
opentitan_functest(
   name = "new_test",
   srcs = ["example_test.c"],
   deps = [
       "@//sw/device/lib/testing/test_framework:ottf_main",
   ],
)
Only when new_test name defined in BUILD.bazel under $REPO_TOP/sw/device/tests/closed_source/ - the  build and run of new_test in my local folder passed successfully

**Comment by timothytrippel** [comment]
I think I have solved your issue, but please read through / perform the steps below to verify this works you. If so, please close the issue, otherwise I can hop on a call to help you resolve this more quickly.

# TL;DR
Add 
```
  exports: [
    {MANUFACTURER_HOOKS_DIR: "/path/to/closed_source_dir"}
  ]
```
to `hw/top_earlgrey/dv/chip_sim_cfg.hjson` instead of setting the`sw_build_opts` to `["--action_env=MANUFACTURER_HOOKS_DIR=/path/to/closed_source_dir"]`

# Complete Steps to Resolve
To reproduce / resolve your issue, I did the following:

1. Created a separate directory on my system for closed source tests by running the following from `$REPO_TOP`: `cp -r sw/device/tests/closed_source /usr/local/google/home/ttrippel/closed_tests`
2. added a new test in the `/usr/local/google/home/ttrippel/closed_tests` directory by copying/renaming the existing `example_test.c` file: `cp /usr/local/google/home/ttrippel/closed_tests/example_test.c /usr/local/google/home/ttrippel/closed_tests/new_test.c`
3. updated the `/usr/local/google/home/ttrippel/closed_tests/BUILD.bazel` file to add the following target:
```
opentitan_functest(
    name = "new_test",
    srcs = ["new_test.c"],
    deps = [
        "@//sw/device/lib/testing/test_framework:ottf_main",
    ],
)
```
4. Note, I **_DID NOT_** modify the open source `BUILD.bazel` file in `$REPO_TOP/sw/device/tests/closed_source/`
5. Ran the test on the FPGA with: `MANUFACTURER_HOOKS_DIR=/usr/local/google/home/ttrippel/closed_tests bazel test @manufacturer_test_hooks//:new_test_fpga_cw310 --test_output=streamed --cache_test_results=no` and this was the output:
<img width="976" alt="Screen Shot 2022-08-22 at 12 30 05 PM" src="https://user-images.githubusercontent.com/5633066/186003151-88d176ee-30b1-43da-9908-a535830e78dc.png">

6. Added the following to the `$REPO_TOP/hw/top_earlgrey/dv/chip_sim_cfg.hjson` file to enable testing this in DV sim:
```
{
  name: chip_sw_example_manufacturer_new
  uvm_test_seq: chip_sw_base_vseq
  sw_images: ["@manufacturer_test_hooks//:new_test:1"]
  en_run_modes: ["sw_test_mode_test_rom"]
}
```
and added 
```
  exports: [
    {MANUFACTURER_HOOKS_DIR: "/usr/local/google/home/ttrippel/closed_tests"}
  ]
```
right after [line 105](https://cs.opensource.google/opentitan/opentitan/+/master:hw/top_earlgrey/dv/chip_sim_cfg.hjson;drc=feb9da73235fc52c099904f4645905ac34560e87;l=106)
7. ran `./util/dvsim/dvsim.py hw/top_earlgrey/dv/chip_sim_cfg.hjson -i chip_sw_example_manufacturer_new --local` to see if Bazel could still build the test SW when invoked by dvsim, and this was the output of the `dvsim.py` produced `run.log` file (**_which indicated a successful build_**):

<img width="1431" alt="Screen Shot 2022-08-22 at 2 08 38 PM" src="https://user-images.githubusercontent.com/5633066/186019430-b9a5fc88-47b6-471a-98f6-571076acaed6.png">
